File: modules/publix/app/session/batch/akka/actors/BatchActionMsgBuilder.java
package session.batch.akka.actors;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.google.common.base.Strings;

import daos.common.BatchDao;
import models.common.Batch;
import play.Logger;
import play.Logger.ALogger;
import play.db.jpa.JPAApi;
import play.libs.Json;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg.BatchAction;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg.TellWhom;

/**
 * @author Kristian Lange (2017)
 */
@Singleton
public class BatchActionMsgBuilder {

	private static final ALogger LOGGER = Logger
			.of(BatchActionMsgBuilder.class);

	private final JPAApi jpa;

	private final BatchDao batchDao;

	@Inject
	public BatchActionMsgBuilder(JPAApi jpa, BatchDao batchDao) {
		this.jpa = jpa;
		this.batchDao = batchDao;
	}

	public BatchActionMsg buildSimple(Batch batch, BatchAction action,
			TellWhom tellWhom) {
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, action.toString());
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

	/**
	 * Sends a batch action message with the current batch session data and
	 * version to all members.
	 */
	public BatchActionMsg buildSessionPatch(Batch batch,
			JsonNode batchSessionPatchNode, TellWhom tellWhom) {
		LOGGER.debug(".buildSessionPatch: batchId {}, studyResultId {}",
				batch.getId());
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, BatchAction.SESSION.toString());
		objectNode.set(BatchActionMsg.BATCH_SESSION_PATCHES,
				batchSessionPatchNode);
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

	/**
	 * Sends a TODO
	 */
	public BatchActionMsg buildWithSession(long batchId, BatchAction action,
			TellWhom tellWhom) {
		LOGGER.debug(".buildWithSession: batchId {}, action {}, tellWhom {}",
				batchId, action, tellWhom.name());
		return jpa.withTransaction(() -> {
			Batch batch = batchDao.findById(batchId);
			if (batch != null) {
				return buildSessionActionMsg(batch, action, tellWhom);
			} else {
				String errorMsg = "Couldn't find batch with ID " + batchId
						+ " in database.";
				return buildErrorActionMsg(errorMsg, TellWhom.SENDER_ONLY);
			}
		});
	}

	/**
	 * Creates a BatchActionMsg. The BatchActionMsg includes a whole bunch of
	 * data including the action, all currently open channels, the group session
	 * version, and the group session data.
	 * 
	 * @param studyResultId
	 *            Which group member initiated this action
	 * @param action
	 *            The action of the GroupActionMsg
	 * @param GroupResult
	 *            The GroupResult of this group
	 */
	private BatchActionMsg buildSessionActionMsg(Batch batch,
			BatchAction action, TellWhom tellWhom) {
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, action.toString());
		if (Strings.isNullOrEmpty(batch.getBatchSessionData())) {
			objectNode.put(BatchActionMsg.BATCH_SESSION_DATA, "{}");
		} else {
			objectNode.put(BatchActionMsg.BATCH_SESSION_DATA,
					batch.getBatchSessionData());
		}
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

	public BatchActionMsg buildError(long batchId, String errorMsg,
			TellWhom tellWhom) {
		return buildErrorActionMsg(errorMsg, tellWhom);
	}

	/**
	 * Creates an ERROR group action message.
	 */
	private BatchActionMsg buildErrorActionMsg(String errorMsg,
			TellWhom tellWhom) {
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, BatchAction.ERROR.toString());
		objectNode.put(BatchActionMsg.ERROR_MSG, errorMsg);
		return new BatchActionMsg(objectNode, tellWhom);
	}

}

File: modules/session/app/batch/BatchActionMsgBuilder.scala
package batch

import java.io.IOException
import javax.inject.{Inject, Singleton}

import com.google.common.base.Strings
import daos.common.BatchDao
import models.common.Batch
import play.api.Logger
import play.api.libs.json.{JsNumber, JsValue, Json}
import play.db.jpa.JPAApi
import batch.BatchDispatcher.BatchAction.BatchAction
import batch.BatchDispatcher.TellWhom.TellWhom
import batch.BatchDispatcher.{BatchAction, BatchActionJsonKey, BatchMsg, TellWhom}

import scala.compat.java8.FunctionConverters.asJavaSupplier

/**
  * Utility class that builds BatchMsgs. So it mostly handles the JSON creation.
  *
  * @author Kristian Lange (2017)
  */
//noinspection ScalaDeprecation
@Singleton
class BatchActionMsgBuilder @Inject()(jpa: JPAApi, batchDao: BatchDao) {

  private val logger: Logger = Logger(this.getClass)

  /**
    * Creates a simple BatchMsg with an error message
    */
  def buildError(errorMsg: String, tellWhom: TellWhom): BatchMsg = {
    val json = Json.obj(
      BatchActionJsonKey.Action.toString -> BatchAction.Error.toString,
      BatchActionJsonKey.ErrorMsg.toString -> errorMsg)
    BatchMsg(json, tellWhom)
  }

  /**
    * Builds a simple BatchMsg with the action and the session version
    */
  def buildSimple(batch: Batch, action: BatchAction, sessionActionId: Long, tellWhom: TellWhom): BatchMsg = {
    logger.debug(s".buildSimple: batchId ${batch.getId}")
    val json = Json.obj(
      BatchActionJsonKey.Action.toString -> action.toString,
      BatchActionJsonKey.SessionActionId.toString -> JsNumber(BigDecimal(sessionActionId)),
      BatchActionJsonKey.SessionVersion.toString -> JsNumber(BigDecimal(batch.getBatchSessionVersion)))
    BatchMsg(json, tellWhom)
  }

  /**
    * Builds a BatchActionMessage with the batch session patch and version
    */
  def buildSessionPatch(batch: Batch, patches: JsValue, tellWhom: TellWhom): BatchMsg = {
    logger.debug(s".buildSessionPatch: batchId ${batch.getId}")
    val json = Json.obj(
      BatchActionJsonKey.Action.toString -> BatchAction.Session.toString,
      BatchActionJsonKey.SessionPatches.toString -> patches,
      BatchActionJsonKey.SessionVersion.toString -> JsNumber(BigDecimal(batch.getBatchSessionVersion)))
    BatchMsg(json, tellWhom)
  }

  /**
    * Builds a BatchMsg with the current batch session data and version
    */
  def buildSessionData(batchId: Long, action: BatchAction, tellWhom: TellWhom): BatchMsg = {
    jpa.withTransaction(asJavaSupplier(() => {
      logger.debug(s".buildSessionData: batchId $batchId, action $action, tellWhom ${
        tellWhom
          .toString
      }")
      val batch = batchDao.findById(batchId)
      if (batch != null) buildSessionAction(batch, action, tellWhom)
      else buildError(s"Couldn't find batch with ID $batchId in database.", TellWhom.SenderOnly)
    }))
  }

  private def buildSessionAction(batch: Batch, action: BatchAction, tellWhom: TellWhom) = {
    val sessionData =
      try
          if (Strings.isNullOrEmpty(batch.getBatchSessionData)) Json.obj()
          else Json.parse(batch.getBatchSessionData)
      catch {
        case e: IOException =>
          logger.error(s".buildSessionActionMsg: invalid session data in DB - batchId " +
            s"${batch.getId}, batchSessionVersion ${batch.getBatchSessionVersion}, " +
            s"batchSessionData ${batch.getBatchSessionData}, error: ${e.getMessage}")
          Json.obj()
      }

    val json = Json.obj(
      BatchActionJsonKey.Action.toString -> action.toString,
      BatchActionJsonKey.SessionData.toString -> sessionData,
      BatchActionJsonKey.SessionVersion.toString -> JsNumber(BigDecimal(batch
        .getBatchSessionVersion)))

    BatchMsg(json, tellWhom)
  }

}
